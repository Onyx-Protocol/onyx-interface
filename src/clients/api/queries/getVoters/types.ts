import { VotersDetails } from 'types';

export interface GetVotersInput {
  id: string;
  support?: boolean;
  limit?: number;
}

export interface VoterResult {
  id: string;
  proposal: string;
  address: string;
  support: boolean;
  votes: string;
  blockNumber: string;
  blockTimestamp: string;
  transactionHash: string;
}

export type GetVotersOutput = VotersDetails;
