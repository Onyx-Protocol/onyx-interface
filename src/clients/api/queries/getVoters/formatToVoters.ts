import BigNumber from 'bignumber.js';
import { VotersDetails } from 'types';

import indexedVotingSupportNames from 'constants/indexedVotingSupportNames';

import { GetVotersApiResponse } from './types';

const formatToVoter = (payload: GetVotersApiResponse): VotersDetails => ({
  result: payload.data.map(({ address, reason, votes, support, has_voted }) => ({
    address,
    voteWeightWei: new BigNumber(votes),
    reason: reason ?? undefined,
    support: has_voted ? indexedVotingSupportNames[support] : 'NOT_VOTED',
  })),
  sumVotes: new BigNumber(payload.metadata.sumVotes),
});

export default formatToVoter;
