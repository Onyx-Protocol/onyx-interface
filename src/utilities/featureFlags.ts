// import { TOKENS } from 'constants/tokens';

export const DISABLED_TOKENS = [];

export const isAssetEnabled = (assetId: string) => {
  if (assetId) return true;
};
