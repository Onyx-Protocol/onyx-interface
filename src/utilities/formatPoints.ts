export const formatPoint = (point: number, digit?: number) => {
  if (digit) {
    return (+point).toLocaleString('en-US', {
      maximumFractionDigits: digit || 2,
    });
  }

  let [real, imaginary] = point.toString().split('.');
  real = (+real).toLocaleString('en-US');
  imaginary = imaginary ? `.${imaginary}` : '';
  return `${real}${imaginary}`;
};
