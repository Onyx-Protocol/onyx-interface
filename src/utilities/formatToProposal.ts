import BigNumber from 'bignumber.js';
import { Proposal } from 'types';

interface FormatToProposalInput {
  call_datas?: string | null;
  callDatas?: string | null;
  signatures: string;
  targets: string;
  values: string;
  against_votes?: string | null;
  againstVotes?: string | null;
  cancel_block?: number | null;
  cancel_timestamp?: number | null;
  cancel_tx_hash?: string | null;
  cancelBlock?: number | null;
  cancelTimestamp?: number | null;
  cancelTxHash?: string | null;
  canceled: number | boolean;
  created_at?: string | null;
  createdAt?: string | null;
  created_block: number | null;
  created_timestamp: number | null;
  created_tx_hash?: string | null;
  description: string;
  end_block?: number;
  end_timestamp?: number | null;
  end_tx_hash?: string | null;
  endBlock?: number;
  endTimestamp?: number | null;
  endTxHash?: string | null;
  eta: number;
  executed: number | boolean;
  executed_block: number | null;
  executed_timestamp: number | null;
  executed_tx_hash: string | null;
  for_votes?: string | null;
  forVotes?: string | null;
  id: number;
  proposer: string;
  queued_block: number | null;
  queued_timestamp: number | null;
  queued_tx_hash: string | null;
  start_block?: number | null;
  start_timestamp?: number;
  start_tx_hash?: string | null;
  startBlock?: number | null;
  startTimestamp?: number;
  startTxHash?: string | null;
  state:
    | 'Pending'
    | 'Active'
    | 'Canceled'
    | 'Defeated'
    | 'Succeeded'
    | 'Queued'
    | 'Expired'
    | 'Executed';
  // JSON
  updated_at?: string | null;
  updatedAt?: string | null;
  voter_count?: number | null;
  voterCount?: number | null;
}

const createDateFromSecondsTimestamp = (timestampInSeconds: number): Date => {
  const inMilliseconds = timestampInSeconds * 1000;
  return new Date(inMilliseconds);
};

const formatToProposal = ({
  against_votes,
  againstVotes,
  cancel_timestamp,
  created_timestamp,
  description,
  end_block,
  end_timestamp,
  executed_timestamp,
  for_votes,
  forVotes,
  id,
  proposer,
  queued_timestamp,
  start_timestamp,
  state,
  created_tx_hash,
  cancel_tx_hash,
  end_tx_hash,
  executed_tx_hash,
  queued_tx_hash,
  start_tx_hash,
  call_datas,
  callDatas,
  signatures,
  targets,
  values,
}: FormatToProposalInput): Proposal => {
  const endDate = end_timestamp ? createDateFromSecondsTimestamp(end_timestamp) : undefined;

  let descriptionObj: Proposal['description'] = {
    version: 'v2',
    title: '',
    description: '',
    forDescription: '',
    againstDescription: '',
    abstainDescription: '',
  };

  try {
    descriptionObj = JSON.parse(description);
  } catch (err) {
    // Split description in half, delimited by the first instance of a break
    // line symbol (\n). The first half corresponds to the title of the
    // proposal, the second to the description
    const [title, descriptionText] = description.split(/\n(.*)/s);

    // Remove markdown characters from title since it's rendered as plain text
    // on the front end
    const plainTitle = title.replaceAll('*', '').replaceAll('#', '');

    descriptionObj = { version: 'v1', title: plainTitle, description: descriptionText };
  }

  const abstainedVotesWei = new BigNumber(0);
  const againstVotesWei = new BigNumber(against_votes || againstVotes || 0);
  const forVotesWei = new BigNumber(for_votes || forVotes || 0);

  let CallDatas: string[] = [];
  if (call_datas) CallDatas = JSON.parse(JSON.parse(call_datas));
  else if (callDatas) CallDatas = JSON.parse(callDatas);

  let Signatures: string[] = [];
  if (call_datas) Signatures = JSON.parse(JSON.parse(signatures));
  else Signatures = JSON.parse(signatures);

  let Targets: string[] = [];
  if (call_datas) Targets = JSON.parse(JSON.parse(targets));
  else Targets = JSON.parse(targets);

  let Values: string[] = [];
  if (call_datas) Values = JSON.parse(JSON.parse(values));
  else Values = JSON.parse(values);

  const actions = CallDatas.map((item, index) => ({
    callData: item,
    signature: Signatures[index],
    target: Targets[index],
    value: Values[index],
  }));

  const proposal: Proposal = {
    abstainedVotesWei,
    againstVotesWei,
    cancelDate: cancel_timestamp ? createDateFromSecondsTimestamp(cancel_timestamp) : undefined,
    createdDate: created_timestamp ? createDateFromSecondsTimestamp(created_timestamp) : undefined,
    description: descriptionObj,
    endBlock: !end_block ? 0 : end_block,
    endDate,
    executedDate: executed_timestamp
      ? createDateFromSecondsTimestamp(executed_timestamp)
      : undefined,
    forVotesWei,
    id,
    proposer,
    queuedDate: queued_timestamp ? createDateFromSecondsTimestamp(queued_timestamp) : undefined,
    startDate: createDateFromSecondsTimestamp(!start_timestamp ? 0 : start_timestamp),
    state,
    createdTxHash: created_tx_hash ?? undefined,
    cancelTxHash: cancel_tx_hash ?? undefined,
    endTxHash: end_tx_hash ?? undefined,
    executedTxHash: executed_tx_hash ?? undefined,
    queuedTxHash: queued_tx_hash ?? undefined,
    startTxHash: start_tx_hash ?? undefined,
    totalVotesWei: abstainedVotesWei.plus(againstVotesWei).plus(forVotesWei),
    actions: actions || [],
  };

  return proposal;
};

export default formatToProposal;
