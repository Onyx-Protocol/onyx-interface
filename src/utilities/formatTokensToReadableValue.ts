import BigNumber from 'bignumber.js';
import { Token } from 'types';

import PLACEHOLDER_KEY from 'constants/placeholderKey';

import { shortenValueWithSuffix } from './shortenValueWithSuffix';

export const formatTokensToReadableValue = ({
  value,
  token,
  minimizeDecimals = false,
  shortenLargeValue = false,
  addSymbol = true,
  removeDecimals = false,
}: {
  value: BigNumber | undefined;
  token: Token;
  minimizeDecimals?: boolean;
  shortenLargeValue?: boolean;
  addSymbol?: boolean;
  removeDecimals?: boolean;
}) => {
  if (value === undefined) {
    return PLACEHOLDER_KEY;
  }

  let decimalPlaces;
  if (minimizeDecimals) {
    decimalPlaces = 4;
  } else {
    decimalPlaces = token.decimals;
  }

  if (removeDecimals) {
    decimalPlaces = 0;
  }

  let symbolPlacement = '';
  if (addSymbol) {
    symbolPlacement = ` ${token.symbol}`;
  }

  if (shortenLargeValue) {
    return `${shortenValueWithSuffix({
      value,
    })}${symbolPlacement}`;
  }

  return `${value.dp(decimalPlaces).toFormat()}${symbolPlacement}`;
};

export default formatTokensToReadableValue;
