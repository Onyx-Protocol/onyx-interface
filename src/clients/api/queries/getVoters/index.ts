import config from 'config';

import getVotersSubGraph from 'utilities/getVotersSubGraph';

import formatToVoters from './formatToVoters';
import { GetVotersInput, GetVotersOutput } from './types';

export * from './types';

const getVoters = async (input: GetVotersInput): Promise<GetVotersOutput> => {
  const proposalVoters = await getVotersSubGraph(config.chainId, input);

  return formatToVoters(proposalVoters);
};

export default getVoters;
