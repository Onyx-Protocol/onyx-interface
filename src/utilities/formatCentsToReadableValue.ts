/* eslint-disable no-nested-ternary */
import BigNumber from 'bignumber.js';
import { shortenValueWithSuffix } from 'utilities';

import PLACEHOLDER_KEY from 'constants/placeholderKey';

const formatCentsToReadableValue = ({
  value,
  shortenLargeValue = false,
}: {
  value: number | BigNumber | undefined;
  shortenLargeValue?: boolean;
}) => {
  if (value === undefined) {
    return PLACEHOLDER_KEY;
  }

  const wrappedValueDollars = new BigNumber(value).dividedBy(100);

  if (!shortenLargeValue) {
    return `$${
      wrappedValueDollars.gt(0.01)
        ? wrappedValueDollars.toFormat(2)
        : wrappedValueDollars.gt(0.001)
        ? wrappedValueDollars.toFormat(3)
        : wrappedValueDollars.gt(0.0001)
        ? wrappedValueDollars.toFormat(4)
        : wrappedValueDollars.toFormat(5)
    }`;
  }

  // Shorten value
  const shortenedValue = shortenValueWithSuffix({
    value: wrappedValueDollars,
    outputsDollars: true,
  });

  return `$${shortenedValue}`;
};

export default formatCentsToReadableValue;
