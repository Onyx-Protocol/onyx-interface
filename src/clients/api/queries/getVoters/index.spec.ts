import { restService } from 'utilities';

import votersResponse from '__mocks__/api/voters.json';

import getVoters from '.';

jest.mock('utilities/restService');

describe('api/queries/getVoters', () => {
  test('returns proposal', async () => {
    (restService as jest.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: votersResponse,
    }));

    const response = await getVoters({
      id: '1',
    });

    expect(response).toMatchSnapshot();
  });
});
