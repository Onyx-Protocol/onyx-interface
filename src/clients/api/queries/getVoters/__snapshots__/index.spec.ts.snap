// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/queries/getVoters returns proposal 1`] = `
Object {
  "result": Array [
    Object {
      "address": "0x1b7081943449db22e4d68c5bac45cfc85b44d3a7",
      "reason": undefined,
      "support": "FOR",
      "voteWeightWei": "1.0998195e+26",
    },
  ],
  "sumVotes": "1.0998195e+26",
}
`;
